src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -11715,6 +11735,298 @@ void Test1(object x1, object? y1)
+class C
+{
+    void Test(object? x)
+    {
+        if (x is C)
+        {
+            x.ToString();
+        }
+        else
+        {
+            x.ToString(); // warn
+        }
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (12,13): warning CS8602: Possible dereference of a null reference.
+                //             x.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(12, 13)
+                );
+        }
+
+        [Fact]
+        public void ConditionalBranching_Is_GenericType()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class Base { }
+class C : Base
+{
+    void Test<T>(C? x) where T : Base
+    {
+        if (x is T)
+        {
+            x.ToString();
+        }
+        else
+        {
+            x.ToString(); // warn
+        }
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (13,13): warning CS8602: Possible dereference of a null reference.
+                //             x.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(13, 13)
+                );
+        }
+
+        [Fact]
+        public void ConditionalBranching_IsConstantPattern_Null()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Test(object? x)
+    {
+        if (x is null)
+        {
+            x.ToString(); // warn
+        }
+        else
+        {
+            x.ToString();
+        }
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (8,13): warning CS8602: Possible dereference of a null reference.
+                //             x.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(8, 13)
+                );
+        }
+
+        [Fact]
+        public void ConditionalBranching_IsConstantPattern_NonNull()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Test(object? x)
+    {
+        const string nonNullConstant = ""hello"";
+        if (x is nonNullConstant)
+        {
+            x.ToString();
+        }
+        else
+        {
+            x.ToString(); // warn
+        }
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (13,13): warning CS8602: Possible dereference of a null reference.
+                //             x.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(13, 13)
+                );
+        }
+
+        [Fact]
+        public void ConditionalBranching_IsConstantPattern_NullConstant()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Test(object? x)
+    {
+        const string? nullConstant = null;
+        if (x is nullConstant)
```

Comment:
>nullConstant [](start = 17, length = 12)

Consider a behavior verification test of an `is` comparison to a nonconstant.

Reply:
That's an existing error ([constant expected](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQDk4BstWFiAJQkiQDsBjCAFQE8AHFAGhhAEssmATEAagB8AAgAYABIICMAbgCwAKEEBmcQCZRAYVEBveaL3jlggCyiAsgAoAlNt367Ae2AArCBRij7ogLyiy2LACEsnJ2dur2ZMj2BAB0AOoIbDAQ5p5syB6WwXYAvvI5QA=)). The `is` operator works with a type, a declaration pattern or a constant pattern.